### DISCUSSION TITLE
Use Dompdf as an alternative PDF renderer

### DISCUSSION THREAD

**Discussion by alexleach** [kimai/kimai#3892] [Plugins]
### Describe the problem

Hello,

I'd like to propose adding an alternative pdf renderer to mpdf: dompdf. From the [mpdf homepage](https://mpdf.github.io/):-

> mPDF as a whole is a quite dated software. Nowadays, better alternatives are available, albeit not written in PHP.

[Dompdf](https://dompdf.github.io/) is an alternative with support for 'most CSS2.1 and a few CSS3 properties'. It has better support for CSS floating blocks for instance, and makes a powerful choice of alternative PDF renderer.

### Describe the solution you'd like

Adding a new class implementation of the `HtmlToPdfConverter` interface, which uses Dompdf instead of Mpdf is fairly trivial (which I think is a testament to how well-designed kimai is!). I [did this yesterday](https://github.com/alexleach/kimai/commit/6f95a5cefe6ee067d2ed4c7b2b812defb2400c2e) and after adding a few lines to [config/services.yaml](https://github.com/alexleach/kimai/commit/6f95a5cefe6ee067d2ed4c7b2b812defb2400c2e#diff-c47b03734cd2b4337b345eeba955637ba790d51fd98979f6d366d4acbdb104e0) can now use Dompdf instead of Mpdf. However, plumbing it in to the Kimai software architecture, using Mpdf as default (to not break the current invoice templates) and allowing Dompdf to be selected in a configuration file and/or environment variable is where I fall flat.

My knowledge of `php`, `composer`, `symfony` etc., and how they work together is trivial at best, and I don't otherwise have a need to become an expert in using them (although I would like kimai to authenticate against an OIDC endpoint, and was going to look into that). But I'd like to use Dompdf, as it appears to have better handling of css 'float' descriptors for my invoice templates, and makes very compact pdfs (when font-subsetting is enabled).

So, I'm not really sure where to take my fork from here. It's got the functionality that I need now, but none of the configurability that I'd expect if I was an end-user making a decision on which Pdf converter to use. And I'd like to keep using an kimai from docker hub, as opposed to a git fork I don't have the intention of maintaining and updating regularly.

Is someone able to take the [DompdfConverter.php](https://github.com/alexleach/kimai/blob/dompdf/src/Pdf/DompdfConverter.php) file I created (obviously improve if and as necessary), and properly plumb it into kimai?

Thanks and kind regards,
Alex

### Describe alternatives you've considered

Maybe it would be better off as a plugin? 🤷 
Really not something I was planning on spending time doing...

### Screenshots

A pdf of some dev / demo data, created with the Dompdf backend and just using the Default Invoice template.

<img width="780" alt="Screenshot 2023-02-24 at 08 55 45" src="https://user-images.githubusercontent.com/375981/221135707-e475c8ec-fd04-4b3f-8b5e-487fe4b7a3be.png">

**Comment by kevinpapst** [comment]
Interesting suggestion 👍  thanks for reaching out!
And yes, 2.5kB is quite impressive, agreed!

> mPDF as a whole is a quite dated software. Nowadays, better alternatives are available, albeit not written in PHP.

That is true, but mPDF does a great job and the problem that I have as maintainer: DomPDF has way more requirements, which makes it hard to ship as default solution. Besides compatibility issues with existing layouts.

But being able to select another rendering engine is a good idea! I'll look into it when I find time.

On a side note: what kind of invoice layout do you have, that does not work with mPDF? Can you share that (if not public via email)?

**Comment by alexleach** [comment]
Thanks for the quick response 😄 

> And yes, 2.5kB is quite impressive, agreed!

I later noticed that it is missing the header and footer, due to the use of `<htmlpageheader>` and `<htmlpagefooter>` tags in https://github.com/kimai/kimai/blob/main/templates/invoice/renderer/default.pdf.twig. Replacing those with `<header>` and `<footer>` tags, respectively gets them showing (although I have no clue actually how those `htmlpage(header|footer)` tags get processed by mpdf, or if it's kimai that handles them). But even with the headers and footers in, dompdf makes a 2.74kB file, vs mpdf's 30.65kB, for _almost_ the same content. So yeah, a >90% size reduction isn't bad!

> DomPDF has way more requirements, which makes it hard to ship as default solution. Besides compatibility issues with existing layouts.

Okay, fair point. I wasn't aware of the difference in number of requirements. And agreed, dompdf couldn't be made the default while the existing templates aren't known to work with it. So it probably is better suited as a plugin then? I guess there's no point in shipping releases with two separate pdf rendering plugins, with all their respective dependencies. I'm totally overwhelmed by all the documentation though. It would take me days or weeks to get familiar enough with all the various libraries, frameworks and intricacies to put together a PR for this! So, thanks for agreeing to look into it! 😄 

Re: layout. I started adjusting the [din-5008 invoice template](https://github.com/kimai/invoice-templates/tree/main/din5008-invoice) to suit my needs, which is still a work in progress. It appeared that `float: left;` wasn't working as I was expecting (and noted here https://mpdf.github.io/what-else-can-i-do/floating-blocks.html). I wanted the first page info to `float`, as opposed to having `absolute` positions. I'll push it to [my dompdf branch](https://github.com/alexleach/kimai/tree/dompdf) when I get it in a decent state, and will refer back to this issue in the commit message (so you should get a notification 😉 ).

**Comment by kevinpapst** [comment]
The special header/footer comment blocks are processed by mpdf, because they need to be applied to every page.

Does dompdf has a similar feature? I'd say: there must be. Question is: how do you define header&footer with it?

When it comes to invoice templates, I would argue that it is better to remember how we did it back in 2000: create your layout with tables. Yes, it sucks, but it works^^

**Comment by alexleach** [comment]
The repetition of headers and footers just seems to work...

After copying [`templates/invoice/renderer/default.pdf.twig`](https://github.com/kimai/kimai/blob/main/templates/invoice/renderer/default.pdf.twig) to `var/invoices/default-copy.pdf.twig`, I can get the headers and footers repeating with some minor changes to the header and footer elements, bringing them out of the `<--mpdf` comment blocks, and adding some css style positioning...

```php
<!DOCTYPE html>
{%- set fallback = app.request is not null ? app.request.locale : 'en' -%}
{%- set language = model.template.language|default(fallback) -%}
{%- set isDecimal = true -%}
{%- set currency = model.currency -%}
{%- set project = model.query.project -%}
<html lang="{{ language }}">
<head>
    <meta charset="UTF-8">
    <title>{% block title %}{{ model.invoiceNumber }}-{{ model.customer.company|default(model.customer.name)|u.snake }}{% endblock %}</title>
    <style>
        {{ encore_entry_css_source('invoice-pdf')|raw }}
header {
    position: fixed;
    top: -1cm;
}
footer {
    position: fixed;
    bottom: 0cm;
}

    </style>
</head>
<body>
<header name="header">
    <table class="header">
        <tr>
            <td class="title">
                {{ model.template.title }}
            </td>
            <td class="date text-right">
                {{ 'date'|trans }}: {{ model.invoiceDate|date_short }}
            </td>
        </tr>
    </table>
</header>
<footer name="footer">
    <table class="footer">
        <tr>
            <td>
                <strong>{{ 'contact'|trans }}</strong>:
                {{ model.template.contact|nl2str(' – ') }}
                <br>
                <strong>{{ 'invoice_bank_account'|trans }}</strong>:
                {{ model.template.paymentDetails|nl2str(' – ') }}
            </td>
            <td align="right">
                {{ 'export.page_of'|trans({'%page%': '{PAGENO}', '%pages%': '{nb}'}) }}
            </td>
        </tr>
    </table>
</footer>
<main>
[...]
</main>
</body>
</html>
```

Haha, yeah I'm not a fan of tables, but agreed, they do work. Writing email templates requires a ridiculous number of tables, too! 🙄 

It was more the absolute positioning of the `invoice_address` and `invoice_details` divs on the first page of the din-5008 template that I was trying to move around and tidy up though. I might have left them alone if they were in a table to be honest...

**Comment by alexleach** [comment]
Hi, I hope you had a good weekend. I (sadly) spent most of it converting this dompdf code into a plugin. Took me pretty much two full days to figure out how to get `HtmlToPdfConverter` to alias my dompdf converter class, instead of `App\Pdf\MPdfConverter` [[1]](#ref1).

I've published it to https://github.com/alexleach/DompdfRendererBundle/, if you'd like to check it out 🙂 

[[1]](#ref1): The Extension class had to implement `CompilerPassInterface`, with a `process` method that sets the alias. See https://github.com/alexleach/DompdfRendererBundle/blob/main/DependencyInjection/DompdfRendererExtension.php.
